### PR TITLE
[Backends/2] Migrate from backendkind to backend statically registered name part 2.

### PR DESCRIPTION
**Pull request by rdzhabarov**
Summary:
Migrate most of the places where BackendKind used to backend name.
Most of the changes are purely mechanical.

Documentation:
updated

Test Plan:
CI

cc: @gcatron

### CODE REVIEW

File: tests/unittests/QuantizationTest.cpp [hunk 1]
```diff
@@ -2221,7 +2215,7 @@ TEST(Quantization, CheckAssertQuantization) {
 /// Check that we can quantize nodes that have some quantized outputs as unused,
 /// e.g. a TopK node where values is unused but indices is.
 TEST(Quantization, QuantizationZeroUsersResult) {
-  ExecutionEngine EE{BackendKind::Interpreter};
+  ExecutionEngine EE{"Interpreter"};
```
Comment (opti-mix) [inline tests/unittests/QuantizationTest.cpp]: Why do you leave `"Interpreter"` here, but removed it in many other places?

File: tests/unittests/QuantizationTest.cpp [hunk 2]
```diff
@@ -2158,7 +2152,7 @@ TEST(Quantization, TestProfileQuantizationOfLoweredFCRowwise) {
 
 /// Check that asserting quantization for the quantizer works as expected.
 TEST(Quantization, CheckAssertQuantization) {
-  ExecutionEngine EE{BackendKind::Interpreter};
+  ExecutionEngine EE{"Interpreter"};
```
Comment (opti-mix) [inline tests/unittests/QuantizationTest.cpp]: Why do you leave "Interpreter" here, but removed it in many other places?

File: tests/unittests/QuantizationTest.cpp [hunk 3]
```diff
@@ -1962,7 +1956,7 @@ static NodeClass *findNodeKindOrReturnNull(Function *F) {
 template <class BackendClass>
 static void testProfileQuantizationOfFC(bool expectLoweredFC,
                                         bool rowwiseQuantizeFC) {
-  ExecutionEngine profileEE(BackendKind::Interpreter);
+  ExecutionEngine profileEE("Interpreter");
```
Comment (opti-mix) [inline tests/unittests/QuantizationTest.cpp]: Why do you leave "Interpreter" here, but removed it in many other places?

File: tests/unittests/QuantizationTest.cpp [hunk 4]
```diff
@@ -622,7 +616,7 @@ TEST(Quantization, quantizeReLU) {
 /// are implemented as IntLookupTables, because the Interpreter only supports
 /// them as such.
 TEST(Quantization, quantizeLookupTables) {
-  ExecutionEngine EE{BackendKind::Interpreter};
+  ExecutionEngine EE{"Interpreter"};
```
Comment (opti-mix) [inline tests/unittests/QuantizationTest.cpp]: Why do you leave "Interpreter" here, but removed it in many other places?

File: tests/unittests/QuantizationTest.cpp (hunk 3 above)
Comment (rdzhabarov) [reply tests/unittests/QuantizationTest.cpp]: Yeah, I'll remove in all. No particular reason.

**Comment by rdzhabarov** [comment]
> Question: Why do you use std::string for storing the backend name (e.g. in the partitioner), but StringRef in some other cases? How do you decide which type to use? Can we use e.g. StringRef almost everywhere?

StringRef for passing. I think storage is std::string everywhere.

File: lib/Partitioner/Partitioner.cpp [hunk 5]
```diff
@@ -69,13 +69,13 @@ void Partitioner::dumpDAG(llvm::StringRef dotFilename) const {
     for (size_t i = 0; i < node->children.size(); i++) {
       auto child = node->children[i];
       DescriptionBuilder db(child->name.c_str());
-      const std::string &backendName = backendMap_.at(child->backendKind).name;
-      db.addParam("BackendKind", backendName);
+      auto backendName = "TEST"; // backendMap_[child->backendName].name;
```
Comment (rdzhabarov) [inline lib/Partitioner/Partitioner.cpp]: TODO: FIX THIS.

**Comment by rdzhabarov** [comment]
@opti-mix yeah, all those cases where std::string is passed in tests/unittests/BackendTestUtils.h is a mistake. I'll fix it!

**Comment by opti-mix** [comment]
@rdzhabarov 

> yeah, all those cases where std::string is passed in tests/unittests/BackendTestUtils.h is a mistake. I'll fix it!

OK. Just be careful with the runtime-related things like partitioner/devicemeanager/etc. I remember seeing some cases, where one would e.g. hold a reference to a function before calling them and then use it after the call. The problem was that this function was gone already due to the changes introduced by the partitioner ;-) As a result you'd get a dangling reference and all the usual problems with it. I hope backends do not disappear in the same way ;-)

**Comment by rdzhabarov** [comment]
doc update (if any) and comment update is pending before merging.

**Comment by rdzhabarov** [comment]
Many files touched, almost every commit brings conflict or failure. I'd like to merge this sooner.

**Comment by rdzhabarov** [comment]
Since there are many moving pieces, many potential conflicts and mechanical nature of the change
I'll land this once fbcode build passes.

Happy to address any further comments in the follow ups.

**Pull request closed without merging**